tracing-appender/src/non_blocking.rs

```diff
@@ -267,3 +267,139 @@ impl Drop for WorkerGuard {
         }
     }
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use std::sync::mpsc;
+    use std::thread;
+    use std::time::Duration;
+
+    struct MockWriter {
+        tx: mpsc::SyncSender<String>,
+    }
+
+    impl MockWriter {
+        fn new(capacity: usize) -> (Self, mpsc::Receiver<String>) {
+            let (tx, rx) = mpsc::sync_channel(capacity);
+            (Self { tx }, rx)
+        }
+    }
+
+    impl std::io::Write for MockWriter {
+        fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+            let buf_len = buf.len();
+            let _ = self.tx.send(String::from_utf8_lossy(buf).to_string());
+            Ok(buf_len)
+        }
+
+        fn flush(&mut self) -> std::io::Result<()> {
+            Ok(())
+        }
+    }
+
+    #[test]
+    fn backpressure_exerted() {
+        let (mock_writer, rx) = MockWriter::new(1);
+
+        let (mut non_blocking, _guard) = self::NonBlockingBuilder::default()
+            .lossy(false)
+            .buffered_lines_limit(1)
+            .finish(mock_writer);
+
+        let error_count = non_blocking.error_counter();
+
+        non_blocking.write_all(b"Hello").expect("Failed to write");
+        assert_eq!(0, error_count.load(Ordering::Acquire));
+
+        let handle = thread::spawn(move || {
+            non_blocking.write_all(b", World").expect("Failed to write");
+        });
+
+        // Sleep a little to ensure previously spawned thread get's blocked on write.
+        thread::sleep(Duration::from_millis(100));
+        // We should not drop logs when blocked.
+        assert_eq!(0, error_count.load(Ordering::Acquire));
+
+        // Read the first message to unblock sender.
+        let mut line = rx.recv().unwrap();
+        assert_eq!(line, "Hello");
+
+        // Wait for thread to finish.
+        let _ = handle.join();
```

Comment:
we should probably unwrap this to ensure that the spawned thread doesn't panic
```suggestion
        let _ = handle.join().expect("thread should not panic");
```